### ISSUE TITLE
genrule runs before volatile-status.txt is generated

### ISSUE THREAD

**Issue by guoshimin** [bazelbuild/bazel#944]
To reproduce:

Make file /tmp/sleep
```
sleep 5
echo "FOO bar"
```
Make it executable
```
chmod +x /tmp/sleep
```
Make BUILD file
```
genrule(
  name = "foo",
  outs = ["foo.out"],
  cmd = '''
{
  for x in `seq 1 10`; do
    ls -l bazel-out
    sleep 1
  done
} > $@'''
)
```
run bazel build
```
bazel build :foo --workspace_status_command=/tmp/sleep
```

the content of foo.out:
```
total 8
drwxr-xr-x  5 shimin  wheel  170 Feb 19 22:35 local_darwin-fastbuild
-rw-r--r--  1 shimin  wheel   60 Feb 19 22:35 stable-status.txt
total 8
drwxr-xr-x  5 shimin  wheel  170 Feb 19 22:35 local_darwin-fastbuild
-rw-r--r--  1 shimin  wheel   60 Feb 19 22:35 stable-status.txt
total 8
drwxr-xr-x  5 shimin  wheel  170 Feb 19 22:35 local_darwin-fastbuild
-rw-r--r--  1 shimin  wheel   60 Feb 19 22:35 stable-status.txt
total 8
drwxr-xr-x  5 shimin  wheel  170 Feb 19 22:35 local_darwin-fastbuild
-rw-r--r--  1 shimin  wheel   60 Feb 19 22:35 stable-status.txt
total 8
drwxr-xr-x  5 shimin  wheel  170 Feb 19 22:35 local_darwin-fastbuild
-rw-r--r--  1 shimin  wheel   60 Feb 19 22:35 stable-status.txt
total 16
drwxr-xr-x  5 shimin  wheel  170 Feb 19 22:35 local_darwin-fastbuild
-r-xr-xr-x  1 shimin  wheel   60 Feb 19 22:35 stable-status.txt
-r-xr-xr-x  1 shimin  wheel   38 Feb 19 22:35 volatile-status.txt
total 16
drwxr-xr-x  5 shimin  wheel  170 Feb 19 22:35 local_darwin-fastbuild
-r-xr-xr-x  1 shimin  wheel   60 Feb 19 22:35 stable-status.txt
-r-xr-xr-x  1 shimin  wheel   38 Feb 19 22:35 volatile-status.txt
total 16
drwxr-xr-x  5 shimin  wheel  170 Feb 19 22:35 local_darwin-fastbuild
-r-xr-xr-x  1 shimin  wheel   60 Feb 19 22:35 stable-status.txt
-r-xr-xr-x  1 shimin  wheel   38 Feb 19 22:35 volatile-status.txt
total 16
drwxr-xr-x  5 shimin  wheel  170 Feb 19 22:35 local_darwin-fastbuild
-r-xr-xr-x  1 shimin  wheel   60 Feb 19 22:35 stable-status.txt
-r-xr-xr-x  1 shimin  wheel   38 Feb 19 22:35 volatile-status.txt
total 16
drwxr-xr-x  5 shimin  wheel  170 Feb 19 22:35 local_darwin-fastbuild
-r-xr-xr-x  1 shimin  wheel   60 Feb 19 22:35 stable-status.txt
-r-xr-xr-x  1 shimin  wheel   38 Feb 19 22:35 volatile-status.txt
```

**Comment by bsilver8192** [comment]
That's because the genrule doesn't actually have a dependency on volatile-status.txt, which also means it won't be re-run if the workspace status information changes. If you build with sandboxing enabled, the script can't see the files at all. Here's what foo.out says on my system (using sandboxing):
```
total 4
drwxr-xr-x 3 nobody nogroup 4096 Feb 20 02:55 local_linux-fastbuild
total 4
drwxr-xr-x 3 nobody nogroup 4096 Feb 20 02:55 local_linux-fastbuild
total 4
drwxr-xr-x 3 nobody nogroup 4096 Feb 20 02:55 local_linux-fastbuild
total 4
drwxr-xr-x 3 nobody nogroup 4096 Feb 20 02:55 local_linux-fastbuild
total 4
drwxr-xr-x 3 nobody nogroup 4096 Feb 20 02:55 local_linux-fastbuild
total 4
drwxr-xr-x 3 nobody nogroup 4096 Feb 20 02:55 local_linux-fastbuild
total 4
drwxr-xr-x 3 nobody nogroup 4096 Feb 20 02:55 local_linux-fastbuild
total 4
drwxr-xr-x 3 nobody nogroup 4096 Feb 20 02:55 local_linux-fastbuild
total 4
drwxr-xr-x 3 nobody nogroup 4096 Feb 20 02:55 local_linux-fastbuild
total 4
drwxr-xr-x 3 nobody nogroup 4096 Feb 20 02:55 local_linux-fastbuild
```

**Comment by guoshimin** [comment]
So what should I do if I want my genrule to have access to
{volatile,stable}-status. txt?

**Comment by damienmg** [comment]
Add the stamp argument, we now set --nostamp, making genrule non stamped by
default. Having stamp = 1 as an attribute of the genrule should do the trick

**Comment by guoshimin** [comment]
That's exactly what I need. Thanks!

**Comment by jxramos** [comment]
Is this `stamp` attribute still available? It's not showing up in the documentation https://bazel.build/reference/be/general#genrule

**Comment by dfr** [comment]
It seems to be available in bazel 5.3.0 but it would be nice if the docs mention it

**Comment by jxramos** [comment]
Confirmed it's definitely in the bazel rule, if I comment out `stamp` I get a failure in my genrule with file not found while uncommenting it out generates the output file as expected
```py
genrule(
    name = "gen_metadata",
    outs = ["metadata.py"],
    cmd = 'cat bazel-out/volatile-status.txt >> $@',
    #stamp = 1,
    visibility = ["//visibility:public"],
)

```


```
ERROR: /home/USERX/workspaceY/proj_foo/BUILD:103:8: Executing genrule //proj_foo/:gen_metadata failed: (Exit 1): bash failed: error executing command (from target //proj_foo/:gen_metadata) /bin/bash -c ... (remaining 1 argument skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
cat: bazel-out/volatile-status.txt: No such file or directory
```
